{
  "cells": [
    {
      "cell_type": "markdown",
      "id": "a8b70b52-1337-4b9a-b248-02afe3b0d4b9",
      "metadata": {
        "id": "a8b70b52-1337-4b9a-b248-02afe3b0d4b9"
      },
      "source": [
        "## Polygenic Scores for Insecticide Resistance Surveillance\n",
        "\n",
        "- [Introductory slides](https://docs.google.com/presentation/d/1qpefQbitedxnI68M82BNK2bZGOC6cPDSXSumE1DpMPY/edit?usp=sharing)\n",
        "- [Git Repo](https://github.com/LSTM-VIGG/pgs)\n",
        "\n",
        "In this workshop, we will explore how **polygenic scores (PGS)** can be used to quantify and track **insecticide resistance** in malaria vector populations.    \n",
        "\n",
        "We will work with MalariaGEN-derived genomic data (`malariagen_data`) and use effect size estimates from a recent targeted amplicon sequencing study using the Ag-vampIR + AmpSeeker platform ([Nagi et al., 2025, *bioRxiv*](https://www.biorxiv.org/content/10.1101/2025.02.14.637727v2)).\n",
        "\n",
        "This notebook will guide you through:\n",
        "- Understanding and interpreting polygenic scores in a vector genomics context  \n",
        "- Using effect size data (odds ratios) to compute polygenic scores  \n",
        "- Applying PGS to individual and population (cohort) mosquito genotype datasets  "
      ]
    },
    {
      "cell_type": "markdown",
      "id": "194dd644-f79e-43ae-8976-7c7b8b712861",
      "metadata": {
        "id": "194dd644-f79e-43ae-8976-7c7b8b712861"
      },
      "source": [
        "#### Why Use Polygenic Scores for Resistance Surveillance?\n",
        "\n",
        "Resistance to insecticides is typically polygenic in nature; many loci contribute to the resistant phenotype. However, traditional resistance diagnostics focus on *single* mutations (e.g. Vgsc-995F) which capture only a small fraction of phenotypic variation, and therefore their actual utility for disease control programmes is limited. Advances in whole-genome and targeted sequencing are now making it possible to affordably genotype mosquitoes at many loci across the genome, but to take advantage of this, we need methods that can summarise the phenotypic effect of those genotypes."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "8b2ea6a4-faae-4f78-afb3-7bb3078d6cd3",
      "metadata": {
        "id": "8b2ea6a4-faae-4f78-afb3-7bb3078d6cd3"
      },
      "source": [
        "#### What Are Polygenic Scores?\n",
        "\n",
        "A **polygenic score (PGS)** — sometimes called a genetic risk score (GRS) — is a single numerical value summarizing the *cumulative contribution* of multiple genetic variants to a particular phenotype.  \n",
        "\n",
        "For insecticide resistance, this phenotype reflects the ability of a mosquito to survive exposure to an insecticide. Each single-nucleotide polymorphism (SNP) or mutation has an **effect size (β, log odds ratio)** that indicates its strength and direction of association with resistance to a given insecticide. To calculate the polygenic score, we simply multiply the effect size by the genotype dosage (0,1,2) and sum across variants.\n",
        "\n",
        "For an individual mosquito $i$, the polygenic score can be represented mathematically as:\n",
        "\n",
        "$PGS_i = \\sum_{j=1}^{m} \\beta_j \\times G_{ij}$\n",
        "\n",
        "where:  \n",
        "- $m$ = number of variants included  \n",
        "- $beta_j$ = estimated *effect size* (e.g. log odds ratio) of variant $j$\n",
        "- $G_{ij}$ = genotype dosage (0, 1, or 2 copies of the resistance allele) for individual $i$\n",
        "\n",
        "This creates a quantitative summary measure that captures small, additive contributions from many loci across the genome — providing a framework for polygenic forms of resistance that cannot be explained by single-point mutations alone."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f2402248-2fab-4a00-b5b5-f784ee4b8110",
      "metadata": {
        "id": "f2402248-2fab-4a00-b5b5-f784ee4b8110"
      },
      "source": [
        "Polygenic scores provide several key advantages:\n",
        "\n",
        "- **Interpretable:** Single easy to interpret value to communicate to policy-makers.\n",
        "- **Standardised:** Enable simple comparisons across studies, populations, and time points.\n",
        "- **Quantitative:** Translate genetic information into a measurable, continuous trait."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "647d4600-7b76-4ea5-8524-012aab736cb3",
      "metadata": {
        "id": "647d4600-7b76-4ea5-8524-012aab736cb3"
      },
      "source": [
        "<figure style=\"text-align: center;\">\n",
        "  <img src=\"https://github.com/LSTM-VIGG/pgs/blob/main/pgs-fig1.png?raw=1\" alt=\"Figure placeholder for PGS schematic\" width=\"600\">\n",
        "  <figcaption><em>Conceptual overview of polygenic score construction for insecticide resistance.</em></figcaption>\n",
        "</figure>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "2998eb34-dcc0-46cf-8581-4c8f1e8b4e08",
      "metadata": {
        "id": "2998eb34-dcc0-46cf-8581-4c8f1e8b4e08"
      },
      "source": [
        "---"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7a3b29bf-adb2-4521-90fe-708c1529f598",
      "metadata": {
        "id": "7a3b29bf-adb2-4521-90fe-708c1529f598"
      },
      "source": [
        "### Workshop Overview\n",
        "\n",
        "In this introductory notebook, we will cover:\n",
        "\n",
        "1. **Data Import** — Loading `malariagen_data` genotypes for target SNPs.  \n",
        "2. **Effect Sizes** — Importing and inspecting effect sizes (β values / log odds ratios) from the Ag-vampIR + AmpSeeker study.  \n",
        "3. **PGS Calculation** — Computing per-individual scores from allele dosages and effect sizes.  \n",
        "4. **Population-Level Summaries** — Aggregating scores to obtain population or cohort-level *mean PGS* values.  \n",
        "5. **Visualization & Interpretation** — Plotting PGS distributions, assessing variation across geography or time."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%pip install malariagen_data -qq"
      ],
      "metadata": {
        "id": "R2kqlG8qdeDe"
      },
      "id": "R2kqlG8qdeDe",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "93c6bec7-4683-443e-a833-d4d244afd2d4",
      "metadata": {
        "id": "93c6bec7-4683-443e-a833-d4d244afd2d4"
      },
      "outputs": [],
      "source": [
        "import malariagen_data\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import allel\n",
        "from tqdm.notebook import tqdm\n",
        "!wget -q https://raw.githubusercontent.com/LSTM-VIGG/pgs/refs/heads/main/pgs_helpers.py\n",
        "import pgs_helpers as pgs\n",
        "\n",
        "ag3 = malariagen_data.Ag3()\n",
        "ag3"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "fa3d3021-6f4b-4a91-be6b-b65cc623d666",
      "metadata": {
        "id": "fa3d3021-6f4b-4a91-be6b-b65cc623d666"
      },
      "source": [
        "#### Load effect size data\n",
        "\n",
        "Lets load the effect size data. The key columns we will need are the `odds_ratio` column and `snp_id`, which tells us the contig, position and alternate allele of the resistance mutation. We can also see other information, such as the p-value and whether this SNP passed FDR control."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f9cd84b1-8b83-41f3-9f2a-a0b41dc8382a",
      "metadata": {
        "id": "f9cd84b1-8b83-41f3-9f2a-a0b41dc8382a"
      },
      "outputs": [],
      "source": [
        "df_effects = pd.read_csv(\"https://raw.githubusercontent.com/LSTM-VIGG/pgs/refs/heads/main/effect-sizes-siaya.csv\", index_col=0)\n",
        "df_effects = df_effects.set_index('snp_id')\n",
        "df_effects.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "494129f4-21ac-4404-8485-f7ed02236fdd",
      "metadata": {
        "id": "494129f4-21ac-4404-8485-f7ed02236fdd"
      },
      "outputs": [],
      "source": [
        "df_effects.shape"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ed926daa-948c-4994-a7cd-33c84e5c6bb4",
      "metadata": {
        "id": "ed926daa-948c-4994-a7cd-33c84e5c6bb4"
      },
      "source": [
        "#### Load genotype data for these SNPs with malariagen_data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "b98d250c-de01-4dcf-91d9-2f97e1b4b869",
      "metadata": {
        "id": "b98d250c-de01-4dcf-91d9-2f97e1b4b869"
      },
      "outputs": [],
      "source": [
        "# define sample sets and query\n",
        "sample_set = 'AG1000G-UG'\n",
        "sample_query = \"taxon == 'gambiae'\"\n",
        "\n",
        "# load sample metadata\n",
        "df_samples = ag3.sample_metadata(sample_set, sample_query)\n",
        "df_samples.shape"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "08257f76",
      "metadata": {
        "id": "08257f76"
      },
      "source": [
        "Lets retrieve snp_calls for every SNP in our effect size data. We will use the `snp_id` column to get the contig, position and alternate allele for each SNP, then call malariagen_data.snp_calls. Because we have relatively few SNPs to retrieve (37), I will loop through each SNP and retrieve them one at a time."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "loc"
      ],
      "metadata": {
        "id": "4RfpPz7pelWO"
      },
      "id": "4RfpPz7pelWO",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "7406491a-0d7b-4532-bcc9-a18af948de15",
      "metadata": {
        "id": "7406491a-0d7b-4532-bcc9-a18af948de15"
      },
      "outputs": [],
      "source": [
        "# setup lists to hold genotype arrays, alt alleles and positions\n",
        "gns = []\n",
        "alts = []\n",
        "poss = []\n",
        "\n",
        "# loop through each SNP in the effect size data\n",
        "for i, row in tqdm(df_effects.iterrows()):\n",
        "\n",
        "    # get the region string from the snp_id\n",
        "    loc = row.name.replace(\"snp_\", \"\").split(\"_\")[0]\n",
        "    region = f\"{loc}-{loc.split(\":\")[1]}\"\n",
        "\n",
        "    # get the snp calls for this region\n",
        "    ds_snps = ag3.snp_calls(region=region, sample_sets=sample_set, sample_query=sample_query)\n",
        "    gn = allel.GenotypeArray(ds_snps['call_genotype'].values)\n",
        "    alt = ds_snps['variant_allele'].values\n",
        "    pos = ds_snps['variant_position'].values\n",
        "\n",
        "    gns.append(allel.GenotypeDaskArray(ds_snps[\"call_genotype\"].data))\n",
        "    alts.append(alt)\n",
        "    poss.append(pos)\n",
        "\n",
        "# concatenate the genotype arrays, alt alleles and positions\n",
        "gn = gns[0].concatenate(gns[1:], axis=0).compute()\n",
        "alts = np.concatenate(alts).astype(str)\n",
        "pos = np.concatenate(poss)\n",
        "samples = ds_snps['sample_id'].values"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "69c49013-6616-4944-a51a-de547c8515eb",
      "metadata": {
        "id": "69c49013-6616-4944-a51a-de547c8515eb"
      },
      "source": [
        "#### Calculate the number of resistance alleles (genotype dosage) for each variant/individual\n",
        "\n",
        "As we are working with relatively few SNPs, I'm going to use pandas dataframes to hold the genotype data. For larger numbers of SNPs (for example, from a GWAS), you would want to use a more efficient data structure, such as scikit-allel's `allel.GenotypeArray`."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "0f1db89a-4ce4-4110-9f0d-110cd09728f1",
      "metadata": {
        "id": "0f1db89a-4ce4-4110-9f0d-110cd09728f1"
      },
      "outputs": [],
      "source": [
        "# create a dataframe with contig, pos, ref, alt information\n",
        "contigs = df_effects.index.str.split(\":\").str.get(0).str.replace(\"snp_\", \"\").values\n",
        "df_var = pd.DataFrame({'contig': contigs, 'pos': pos, 'ref': alts[:, 0].astype(str), 'alt': np.apply_along_axis(lambda x: ','.join(x), axis=1, arr=alts[:, 1:].astype(str))})\n",
        "\n",
        "# create a dataframe with genotypes\n",
        "df_geno = pd.DataFrame(gn.to_gt().astype(str), columns=samples)\n",
        "\n",
        "# combine var_df and df_geno\n",
        "df_geno = pd.concat([df_var, df_geno], axis=1)\n",
        "df_geno.head(3)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "682e5fba",
      "metadata": {
        "id": "682e5fba"
      },
      "source": [
        "As we have some alleles which are multiallelic, we will need to split these into multiple rows, one for each alternate allele. We will then convert the genotype strings (e.g. \"0/1\") into a count of the number of alternate alleles (0, 1 or 2)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "ff95e1d2-918c-4144-80e9-503502b9e043",
      "metadata": {
        "id": "ff95e1d2-918c-4144-80e9-503502b9e043"
      },
      "outputs": [],
      "source": [
        "df_geno = pgs.split_rows_with_multiple_alleles(df_geno, samples)\n",
        "df_geno = pgs.convert_genotype_to_alt_allele_count(df_geno, samples)\n",
        "\n",
        "# create a snp_id column and set as index, and drop unneeded columns\n",
        "df_geno = df_geno.assign(snp_id=lambda x: \"snp_\" + x.contig + \":\" + x.pos.astype(str) + \"_\" + x.alt)\n",
        "df_geno = df_geno.set_index('snp_id')\n",
        "df_geno = df_geno.drop(columns=['contig', 'pos', 'ref', 'alt'])\n",
        "df_geno.head(3)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c9852303-2282-4658-bdeb-579556026650",
      "metadata": {
        "id": "c9852303-2282-4658-bdeb-579556026650"
      },
      "source": [
        "#### Remove SNPs in high LD"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "26c7f523-6ac2-4124-a40d-74dcdc6d6a83",
      "metadata": {
        "id": "26c7f523-6ac2-4124-a40d-74dcdc6d6a83"
      },
      "outputs": [],
      "source": [
        "target_snps = df_effects.index.to_list()\n",
        "df_geno = df_geno.loc[target_snps]\n",
        "\n",
        "retained_snps = pgs.linked(df_geno.T, df_effects)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "retained_snps"
      ],
      "metadata": {
        "id": "BOCGpALUgz0u"
      },
      "id": "BOCGpALUgz0u",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df_geno.shape"
      ],
      "metadata": {
        "id": "uX2h_KLShVfM"
      },
      "id": "uX2h_KLShVfM",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# df_geno.reset_index().query(\"snp_id in @selected_snps\")"
      ],
      "metadata": {
        "id": "5MhWngB4hmdD"
      },
      "id": "5MhWngB4hmdD",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df_geno_unlinked = df_geno[:, retained_snps.tolist()]\n",
        "df_geno_unlinked"
      ],
      "metadata": {
        "id": "81sJtMzNg2UB"
      },
      "id": "81sJtMzNg2UB",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "b234855c",
      "metadata": {
        "id": "b234855c"
      },
      "source": [
        "Now we can select only the snp_ids which are in our effect size data, and calculate PGS."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "7d075e09-95cb-4e2c-a57b-8a566f4002d9",
      "metadata": {
        "id": "7d075e09-95cb-4e2c-a57b-8a566f4002d9"
      },
      "outputs": [],
      "source": [
        "df_pgs = pgs.calculate_pgs(df_geno_unlinked, df_effects, df_samples)\n",
        "df_pgs.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "be150557",
      "metadata": {
        "id": "be150557"
      },
      "source": [
        "Lets look at the distribution of PGS values across individuals."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "2c969b76-b265-48e1-bd95-66a51c8ee716",
      "metadata": {
        "id": "2c969b76-b265-48e1-bd95-66a51c8ee716"
      },
      "outputs": [],
      "source": [
        "df_pgs.pgs.hist()"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.12.8"
    },
    "colab": {
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}